src/crystal/system/win32/file_descriptor.cr

```diff
@@ -274,6 +273,47 @@ module Crystal::System::FileDescriptor
     end
     ret
   end
+
+  private def console_mode?(handle)
+    LibC.GetConsoleMode(handle, out _) != 0
+  end
+
+  private def read_console(handle : LibC::HANDLE, slice : Slice(UInt16)) : Int32
+    if 0 == LibC.ReadConsoleW(handle, slice, slice.size, out units_read, nil)
+      raise IO::Error.from_winerror("ReadConsoleW")
+    end
+    units_read.to_i32!
+  end
+
+  private def read_console(handle : LibC::HANDLE, slice : Bytes) : Int32
+    # Characters between U+0000 and U+FFFF have 1 UTF-16 code unit and up to 3 UTF-8 code unit.
+    # Characters between U+10000 and U+10FFFF have 2 UTF-16 code units and 4 UTF-8 code units.
+    # The following conclusions can therefore be drawn:
+    # 1. N UTF-16 code units correspond to no more than 3*N UTF-8 code units.
+    # 2. We can re-use the last 2/3 of the UTF-8 buffer to receive the UTF-16 data without worrying about data overlap.
+
+    third = slice.size // 3
+    units_buffer = (slice + third).unsafe_slice_of(UInt16)
+
+    # One character may have two UTF-16 code units.
+    raise IO::Error.new("Buffer size too small") if third < 2
```

Comment:
`slice` is not necessarily the internal buffer because one could turn off read buffering at any time. In that case this PR raises an error where none were present before:

```crystal
STDIN.read_buffering = false
b = uninitialized UInt8[4]
STDIN.read(b.to_slice)
```

Reply:
Let's continue the discussion about tiny buffer sizes in #13731